frontends/p4/def_use.cpp

```diff
@@ -46,8 +46,21 @@ StorageLocation* StorageFactory::create(const IR::Type* type, cstring name) cons
         type = typeMap->getTypeType(type, true);  // get the canonical version
         auto st = type->to<IR::Type_StructLike>();
         auto result = new StructLocation(type, name);
+
+        // For header unions we will model all of the valid fields
+        // for all components as a single shared field.  The
+        // reason is that updating one of may change all of the
+        // other ones.
+        StorageLocation* globalValid = nullptr;
+        if (type->is<IR::Type_HeaderUnion>())
+            globalValid = create(IR::Type_Boolean::get(), name + "." + validFieldName);
+
         for (auto f : st->fields) {
-            auto sl = create(f->type, name + "." + f->name);
+            cstring fieldName = name + "." + f->name;
+            auto sl = create(f->type, fieldName);
+            if (globalValid != nullptr)
+                dynamic_cast<StructLocation*>(sl)->replaceField(
```

Comment:
I would guess because the `to` methods currently all return `const` pointers.  We should probably overload them to return non-const pointers when called on non-const objects.